lib/executor/executor.cpp

```diff
@@ -206,5 +207,190 @@ Executor::asyncInvoke(const Runtime::Instance::FunctionInstance *FuncInst,
+
+  //  final file generation
+  AST::Module Mod{};
+  std::vector<Byte> &Magic = Mod.getMagic();
+  Magic.insert(Magic.begin(), {0x00, 0x61, 0x73, 0x6d});
+
+  std::vector<Byte> &Version = Mod.getVersion();
+  Version.insert(Version.begin(), {0x01, 0x00, 0x00, 0x00});
+
+  Mod.getCustomSections().emplace_back(Core);
+  // Mod.getCustomSections().emplace_back(CoreModules);
+  // Mod.getCustomSections().emplace_back(CoreInstances);
+  Mod.getCustomSections().emplace_back(CoreStack);
+  // Mod.getDataSection() = std::move(DataSection);
+  Mod.getMemorySection() = std::move(MemSec);
+  // TODO globals works with DWARF data, is this section needed?
+  //  Mod.getGlobalSection() = std::move(Globals);
+  auto Res = Ser.serializeModule(Mod);
+  std::ofstream File("coredump.wasm", std::ios::out | std::ios::binary);
+  if (File.is_open()) {
+    File.write(reinterpret_cast<const char *>(Res->data()), Res->size());
+    File.close();
+  } else {
+    // TODO change to assuming
+    //  Handle error opening file
+    throw std::ios_base::failure("Failed");
+  }
+}
+
+AST::CustomSection Executor::collectProcessInformation() {
+  spdlog::info("Constructing Core");
+  // AST node
+  AST::CustomSection Core;
+  Core.setName("core");
+
+  // Trying to insert this in leads to wasmgdb
+  // saying it is an unsupported process-type
+  // std::string Name = "a.wasm";
+  // std::vector<Byte> Data(Name.begin(), Name.end());
+  // if (Data.size() != 0) { // case where the module has an empty name field
+  //   Content.insert(Content.end(), Data.begin(), Data.end());
+  // } else {
+  //   Content.insert(Content.begin(), {0x00, 0x00});
+  // }
+  auto &Content = Core.getContent();
+  Content.insert(Content.begin(), {0x00, 0x00});
+
+  return Core;
+}
+
+AST::MemorySection Executor::collectMemory(Runtime::StackManager &StackMgr) {
+  spdlog::info("Collecting memory section");
+
+  AST::MemorySection MemSec;
+  const auto *CurrentInstance = StackMgr.getModule();
+  auto MemInstances = CurrentInstance->getMemoryInstances();
+  std::cout << MemInstances.size() << "\n";
+  std::cout << MemInstances[0]->getMemoryType().getLimit().hasMax() << "\n";
+  std::cout << MemInstances[0]->getMemoryType().getLimit().getMax() << "\n";
+  std::cout << MemInstances[0]->getMemoryType().getLimit().getMin() << "\n";
+  MemInstances[0]->getMemoryType().getLimit();
+  // In the current version of WebAssembly, at most one memory may be defined
+  // or imported in a single module, and all constructs implicitly reference
+  // this memory.
+  //
+  // If hasMax, set Max to that.
+  // Else set whatever Min exists
+
+  return MemSec;
+}
+
+// Data section
+// TODO malformed section
+AST::DataSection Executor::collectDataSection(Runtime::StackManager &StackMgr) {
+  spdlog::info("Collecting Memories");
+
+  AST::DataSection Datasec;
+
+  const auto *CurrentInstance = StackMgr.getModule();
+  auto &DataInstances = CurrentInstance->getOwnedDataInstances();
```

Comment:
IIRC, the data instances should be empty after the instantiation. You may need to re-construct the data instances from the AST. But it's fine to leave a TODO here and implement it later.

Reply:
That is correct, the data instances are empty after instantiation. I don't think wasmgdb relies on the coredump's data section, and rather reconstructs it from the original binary(I think). Even in coredump-rewriter the data section is filled with long runs of zeroes